### PR TITLE
[search] Fuzzy search by category.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

File: base/mem_trie.hpp [hunk 1]
```diff
@@ -27,6 +32,31 @@ class MemTrie
...
 
+  // A read-only iterator. Any modification to the
+  // underlying trie is assumed to invalidate the iterator.
+  class Iterator
+  {
+  public:
+    Iterator(MemTrie::Node const & node) : m_node(node) {}
+
+    template <typename ToDo>
+    void ForEachMove(ToDo && todo) const
```
Comment (ygorshenin) [inline base/mem_trie.hpp]: s/todo/toDo/g

Comment (ygorshenin) [inline base/mem_trie.hpp]: Please, add comments to the public API.

File: base/mem_trie.hpp [hunk 2]
```diff
@@ -27,6 +32,31 @@ class MemTrie
...
+    {
+      for (auto const & move : m_node.m_moves)
+        todo(move.first, Iterator(*move.second));
+    }
+
+    template <typename ToDo>
+    void ForEachInNode(ToDo && todo) const
+    {
+      for (auto const & value : m_node.m_values)
```
Comment (ygorshenin) [inline base/mem_trie.hpp]:
```c++
std::for_each(m_node.m_values.begin(), m_node.m_values.end(), std::forward<ToDo>(toDo));
```
?

File: indexer/categories_holder.hpp [hunk 3]
```diff
@@ -125,6 +125,15 @@ class CategoriesHolder
   /// @returns raw classificator type if it's not localized in categories.txt.
   string GetReadableFeatureType(uint32_t type, int8_t locale) const;
 
+  // Exposes the tries that map category tokens to types.
+  Trie const * GetNameToTypesTrie(int8_t locale) const
```
Comment (ygorshenin) [inline indexer/categories_holder.hpp]: Can we use just a single trie, with first-level edges corresponding to locales?

File: search/utils.hpp [hunk 4]
```diff
@@ -3,31 +3,125 @@
...
+    {
+      auto const * trie = categories.GetNameToTypesTrie(locale);
+      if (trie != nullptr)
+        MatchInTrie(*trie, dfa, std::bind<void>(todo, i, std::placeholders::_1));
+    }
+  }
+}
+
+template <typename IterT1, typename IterT2>
+bool StartsWith(IterT1 beg, IterT1 end, IterT2 begPrefix, IterT2 endPrefix)
```
Comment (ygorshenin) [inline search/utils.hpp]: Could you please move this to base/string_utils, and re-use in StartsWith() version for UniStrings?

File: search/processor.cpp [hunk 5]
```diff
@@ -128,8 +128,8 @@ void SendStatistics(SearchParams const & params, m2::RectD const & viewport, Res
   GetPlatform().GetMarketingService().SendMarketingEvent(marketing::kSearchEmitResultsAndCoords, {});
 }
 
-// Removes all full-token stop words from |params|, unless |params|
-// consists of all such tokens.
+// Removes all full-token stop words from |params|.
+// Does nothing if all tokens in |params| are stop words.
```
Comment (ygorshenin) [inline search/processor.cpp]: What if all tokens except prefix token are stop words?

File: search/utils.hpp [hunk 6]
```diff
@@ -3,31 +3,125 @@
...
 #include "base/string_utils.hpp"
 
+#include <algorithm>
+#include <cctype>
+#include <functional>
+#include <queue>
+
 namespace search
 {
+namespace
```
Comment (ygorshenin) [inline search/utils.hpp]: There's no sense in the anonymous namespace in the header for template function.

File: base/mem_trie.hpp (hunk 1 above)
Comment (mpimenov) [inline base/mem_trie.hpp]: done

File: base/mem_trie.hpp (hunk 2 above)
Comment (mpimenov) [inline base/mem_trie.hpp]: done

File: search/utils.hpp (hunk 6 above)
Comment (mpimenov) [inline search/utils.hpp]: Indeed. Removed.

File: search/processor.cpp (hunk 5 above)
Comment (mpimenov) [inline search/processor.cpp]: Changed to 
```// Does nothing if all tokens in |params| are non-prefix stop words.```

File: search/utils.hpp (hunk 4 above)
Comment (mpimenov) [inline search/utils.hpp]: done
We lost the "len < prefixLen" fast path but I think it's no big deal.

File: indexer/categories_holder.hpp (hunk 3 above)
Comment (mpimenov) [inline indexer/categories_holder.hpp]: I did it in a separate commit. Does it look better now?

**Comment by mpimenov** [comment]
PTAL

File: indexer/categories_holder.hpp [hunk 7]
```diff
@@ -56,7 +56,8 @@ class CategoriesHolder
   Type2CategoryCont m_type2cat;
 
   // Maps locale and category token to the list of corresponding types.
-  map<int8_t, unique_ptr<Trie>> m_name2type;
+  // Locale is treated as a special symbol prepended to the token.
+  unique_ptr<Trie> m_name2type;
```
Comment (ygorshenin) [inline indexer/categories_holder.hpp]: So, do you really need a unique_ptr now?

File: search/utils.hpp [hunk 8]
```diff
@@ -3,31 +3,120 @@
...
+// whose edit distance is close enough to the required token from |slice|.
+template <typename ToDo>
+void ForEachCategoryTypeFuzzy(StringSliceBase const & slice, TLocales const & locales,
+                              CategoriesHolder const & categories, ToDo && todo)
+{
+  using Trie = my::MemTrie<strings::UniString, uint32_t>;
+
+  auto const & trie = categories.GetNameToTypesTrie();
+  auto const & trieRootIt = trie.GetRootIterator();
+  std::set<int8_t> localeSet(locales.begin(), locales.end());
```
Comment (ygorshenin) [inline search/utils.hpp]: Consider to replace std::set by a sorted vector with binary_search.

File: search/utils.hpp [hunk 9]
```diff
@@ -3,31 +3,120 @@
...
+
+  for (size_t i = 0; i < slice.Size(); ++i)
+  {
+    auto const & token = slice.Get(i);
+    auto const & dfa =
```
Comment (ygorshenin) [inline search/utils.hpp]:
```c++
strings::LevenshteinDFA dfa(token, ...);
```

File: search/utils.hpp [hunk 10]
```diff
@@ -3,31 +3,120 @@
...
+        strings::LevenshteinDFA(token, 1 /* prefixCharsToKeep */, GetMaxErrorsForToken(token));
```
Comment (ygorshenin) [inline search/utils.hpp]: At least, leave a TODO here that we build LevenshteinDFA twice for each token (here and in the Geocoder), not quite expensive but still may be optimized.

File: indexer/categories_holder.cpp [hunk 11]
```diff
@@ -203,6 +203,8 @@ void CategoriesHolder::AddCategory(Category & cat, vector<uint32_t> & types)
       auto const locale = synonym.m_locale;
       ASSERT_NOT_EQUAL(locale, kUnsupportedLocaleCode, ());
 
+      auto const localePrefix = String(1, static_cast<strings::UniChar>(locale));
```
Comment (bykoianko) [inline indexer/categories_holder.cpp]: In line comment for 1?

Comment (mpimenov) [inline indexer/categories_holder.cpp]: Do we really need it? The UniString constructor mimics std::string constructor: ```String(numChars, ch)```.

File: search/utils.hpp (hunk 9 above)
Comment (mpimenov) [inline search/utils.hpp]: But it won't be const then!

File: indexer/categories_holder.hpp (hunk 7 above)
Comment (mpimenov) [inline indexer/categories_holder.hpp]: It's easier to implement ```swap``` with unique_ptr.

File: search/utils.hpp (hunk 8 above)
Comment (mpimenov) [inline search/utils.hpp]: done

File: search/utils.hpp (hunk 9 above)
Comment (mpimenov) [inline search/utils.hpp]: done

File: search/utils.hpp (hunk 10 above)
Comment (mpimenov) [inline search/utils.hpp]: done

File: indexer/categories_holder.cpp (hunk 11 above)
Comment (mpimenov) [inline indexer/categories_holder.cpp]: done

**Comment by mpimenov** [comment]
PTAL

File: indexer/categories_holder.hpp (hunk 7 above)
Comment (mpimenov) [inline indexer/categories_holder.hpp]: done

**Pull request merged**